devito/types/equation.py

```diff
@@ -140,10 +142,26 @@ def _flatten(self):
         else:
             return [self]
 
-    @property
+    @cached_property
     def subdomain(self):
         """The SubDomain in which the Eq is defined."""
-        return self._subdomain
+        if self._subdomain is not None:
+            return self._subdomain
+
+        from devito.symbolics.search import retrieve_functions  # noqa
+
+        funcs = retrieve_functions(self)
+        subdomains = {f.grid for f in funcs} - {None}
+        subdomains = {g for g in subdomains if g.is_SubDomain}
+
+        if len(subdomains) == 0:
+            return None
+        elif len(subdomains) == 1:
+            subdomain = subdomains.pop()
+            return subdomain
+        else:
+            raise ValueError("Multiple SubDomains present in Eq, but no iteration"
```

Comment:
I wonder whether there would be value in having any failing-in-purpose-tests to showcase bad use cases like this

Reply:
I think we need something more descriptive, something along the lines of

```
"Multiple SubDomains detected. Provide a SubDomain explicitly (i.e., via `Eq(..., subdomain=...)`) to unambiguously define the `Eq`'s iteration domain."